### ISSUE TITLE
SPIRVReader crashes when translating OpVariable with OpConstantSampler initializer

### ISSUE THREAD

**Issue by aratajew** [KhronosGroup/SPIRV-LLVM-Translator#1316]
Below SPIRV code crashes in SPIRV Reader:
```
%2 = OpTypeSampler
%_ptr_UniformConstant_2 = OpTypePointer UniformConstant %2

%_sampler_init = OpConstantSampler %2 Clamp 0 Nearest
%image_sampler = OpVariable %_ptr_UniformConstant_2 UniformConstant %_sampler_init
```
The crash occurs during translation of `OpVariable`s initializer (`%_sampler_init`):
1. `%image_sampler` is translated to llvm `GlobalVariable`, since it's `OpVariable`:
```llvm
@image_sampler = external addrspace(2) constant %spirv.Sampler addrspace(2)*
```
2. `%_sampler_init` is `OpConstantSampler`, so it should be translated to:
```llvm
call %spirv.Sampler addrspace(2)* @__translate_sampler_initializer(i32 23)
```
But translation crashes because call instruction needs a `BasicBlock` which is `null`, because it's passed from `OpVariable` handling which operates globally.

I'm wondering how such case should be handled. The main problem is that `GlobalVariable` initializer must be `Constant`, but it actually isn't `Constant`, because `OpConstantSampler` is translated to `CallInst`.

**Comment by aratajew** [comment]
@AlexeySachkov, @AlexeySotkin, @svenvh, @MrSidims  I'll be grateful for any insights.

**Comment by svenvh** [comment]
Maybe we should special-case the translation of sampler OpVariables?  Clang doesn't emit GlobalVariables for samplers either, see https://github.com/llvm/llvm-project/blob/919738739a23f3e7a4106c34f0e9aaedd1eb8ecf/clang/lib/CodeGen/CodeGenModule.cpp#L4461

**Comment by aratajew** [comment]
> Maybe we should special-case the translation of sampler OpVariables? Clang doesn't emit GlobalVariables for samplers either, see https://github.com/llvm/llvm-project/blob/919738739a23f3e7a4106c34f0e9aaedd1eb8ecf/clang/lib/CodeGen/CodeGenModule.cpp#L4461

Do you mean to create call to `__translate_sampler_initializer` every time sampler OpVariable is used in a kernel?

**Comment by svenvh** [comment]
Yes, that would match the LLVM IR that Clang generates at least.

**Comment by aratajew** [comment]
Prepared such a change: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1317 It inserts call to `__translate_sampler_initializer` for all uses of `OpVariable`. But frankly saying I can't figure out the proper usage of such `OpVariable`, the kernel I'd like to avoid translation crash for just declares such variable, but doesn't used it at all..

Do you think that we also need a special handling for `OpLoad`/`OpStore` operating on this `OpVariable`? Since currently produced load is illegal according to llvm validation rules:
```llvm
Fails to verify module: loading unsized types is not allowed
  %0 = load %spirv.Sampler, %spirv.Sampler addrspace(2)* %image_sampler, align 8
```

**Comment by AlexeySotkin** [comment]
@aratajew does your SPIR-V module pass `spirv-val`?

**Comment by aratajew** [comment]
> @aratajew does your SPIR-V module pass `spirv-val`?

Yes. It seems that SPIRV specification doesn't restrict `OpVariable` with `OpTypeSampler` type.

**Comment by AlexeySotkin** [comment]
> > @aratajew does your SPIR-V module pass `spirv-val`?
> 
> Yes. It seems that SPIRV specification doesn't restrict `OpVariable` with `OpTypeSampler` type.

Ok

**Comment by aratajew** [comment]
@AlexeySotkin, @svenvh are you ok with merging https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1317 so far to avoid crash on kernels with such variables declared but not used? In the future we may consider implementing support for loads/stores when such cases come up.

**Comment by AlexeySotkin** [comment]
@aratajew could you please add a test to your patch?

**Comment by aratajew** [comment]
> @aratajew could you please add a test to your patch?

Done.
